[HVG-439] Fix / Checkout sign button

styrelius opened the pull request:
<!-- 
If there is one, add the ID of the Jira issue to the Pull Request title.
E.g. "[HVG-123] Make the Pull Request template great again" 
-->

## What?

Fix the design of the `Sign` component, mainly for it to no longer cover a big portion of the `Checkout` sidebar but also just to make it look nicer in general. 🎨 
Changes is this PR entails:
1) Moving the `SignDisclaimer` out of the `Sign` component and put it in its own file. Import it and render it as the last item in the sidebar.
2) Changing the design considerably, for several components, with Manges new designs for the `Checkout` sidebar in mind.
3) Modifying how the `Sign` component is rendered and positioned.
4) Adding logic to get the width of the rendered `CheckoutComponent` minus eventual scroll bar so that the width of `Sign` `Wrapper` can be adjusted to not cover said scrollbar. This is really only an issue with Edge and Explorer since Chrome actually updated this (i.e. removed their scrollbars) during the time I was working on this. If anyone thinks this adds unnecessary complexity for a limited case, please let me know and I will remove it! 
See `src/client/pages/OfferNew/Checkout/index.tsx` row 181-199 for this.
5) General refactoring of the components concerned - moving stuff where it's in my mind more logical to have it, renaming some things etc.


## Why?

<!-- Why are these changes made? -->
This is part of a bigger remake of the signup process, which came about due to the realisation that many potential members/customers (mainly in Norway) are not able to sign up for Hedvig, sometimes because they might not understand how to do it since our UI might not be clear enough. The changes made here belong to those we concluded would be of the more low-hanging fruit type 🍎


<!-- If there is one, add a link to the Jira ticket below. -->
_Referenced ticket: [HVG-439]_


<!-- And, if that makes sense, add screenshots and/or GIF's below -->

## Screenshots and GIF's

**Norwegian checkout with bundled quotes**

<img width="1830" alt="2020-11-24-sign-no" src="https://user-images.githubusercontent.com/42962286/100118686-60347000-2e76-11eb-9a21-05dbbd4426b5.png">

![2020-11-24-sign-no-combo](https://user-images.githubusercontent.com/42962286/100118754-6e828c00-2e76-11eb-8e3c-5dce0b64179d.gif)

**Swedish checkout**

<img width="1830" alt="2020-11-24-sign-se" src="https://user-images.githubusercontent.com/42962286/100119128-d0db8c80-2e76-11eb-86d2-faecd457a1f9.png">


**Starting a BankID sign session in Swedish checkout**

![2020-11-24-sign-se-sign-started](https://user-images.githubusercontent.com/42962286/100119090-c6b98e00-2e76-11eb-8bbe-4e4784361ace.gif)



[HVG-123]: https://hedvig.atlassian.net/browse/HVG-123
[HVG-439]: https://hedvig.atlassian.net/browse/HVG-439

styrelius committed:
Remove TODO's in CheckotContent.tsx and OfferNew/index.tsx since they don't apply anymore.

styrelius committed:
Modify Sign.tsx, Checkout/index.tsx and ChechkoutContent.tsx in order for the Sign component (sign button and disclaimer for terms and GDPR) to be placed at them bottom of checkout sidebar instead of being sticky/fixed.

styrelius committed:
Replace text with spinner on Button in Sign.tsx if loading is true (instead of having the spinner to the right of text).

styrelius committed:
Change name of 'loading' prop passed to Sign component, to 'isLoading'.

styrelius committed:
Tweak styling of sign status in Sign.tsx slightly.

styrelius committed:
Merge branch 'master' into fix/move-checkout-sign-btn

styrelius committed:
Change width and color of Button in Sign.tsx. Fix disabled style in buttons.tsx so that it works with the new purple Sign button color.

styrelius committed:
Hide Intercom button when checkout sidebar is opened. This was previously done for smaller screens but it's actually really annoying in desktop as well, so this means remove the media query check conditional for hiding it.

styrelius committed:
Fix styling of Sign component.
Make the Sign component fixed/sticky again, remove a lot of its padding and change its colors.
Use the outerWrapper ref in Checkout/index.tsx to calculate the width of the Sign Wrapper, since it should be 100% of Checkout sidebar minus scrollbar width. Do this with a new 'wrapperScrollWidth' state and a new useEffect, and pass the value as a prop to Sign.

styrelius committed:
Remove setting 'visibility' to 'hidden' in Sign Wrapper when there is no 'width' value. It's not worth risking that the component is invisible if there's something cempletely off with calculating the width.

styrelius committed:
Make Sign Wrapper 'sticky' instead of 'fixed' since that apparently solved the issue with scrolling when hovering that component. Also move Sign component outside of <InnerWrapper> in Checkout/index.tsx to make the width right.

styrelius committed:
Make 'checkoutWrapperScrollWidth' max-width instead of width in Sign Wrapper. Set width to 100%. Add event listener in Checkout/index.tsx listening to 'resize', in useEffect where 'wrapperScrollWidth' state is set.

styrelius committed:
Move SignDisclaimer component from Sign.tsx into its own file.
Import the new component in Checkout/index and render it below ChekoutContent.
Adjust styling in Sign.tsx.

styrelius committed:
Move SignStatusWrapper from Sign.tsx into SignStatus.tsx. Import colors from brand repo there and tweak some styling.

styrelius committed:
Make some styling tweaks and modifications in Sign.tsx, SignDisclaimer.tsx and Checkout/index.tsx - mostly decreasing padding :)

styrelius committed:
Merge branch 'master' into fix/checkout-sign-btn

gustaveen commented on src/client/pages/OfferNew/Checkout/Sign.tsx:
```diff
@@ -1,69 +1,33 @@
-    justify-content: center;
-  }
+  max-width: ${({ maxWidth }) => (maxWidth ? `${maxWidth}px` : '100%')};
+  padding: ${({ isDesktop }) => (isDesktop ? '1.5rem 2rem' : '1rem')};
```
Maybe it's just me, but can't we just use a good ol' media query instead of sending a js calculated value. *old man shakes fist at computer 😆

styrelius replied:
Haha! 😍 No but my idea was that we'll _centralise_ the media queries later on so we always use the same values everywhere. Like we'll have a nice li'l hook that returns `isDesktop` for example. I guess I made it this way to make it easier to change that. And I actually like a named variable more than a regular media query - to me it's clearer what the media query is for then.

gustaveen replied:
I agree that is nice to have one value as reference. :) In this case we could do like we do in a few other places, set a const at the top of the file `const DESKTOP = 600` or and then we could use it both like so. Then we get the best of both worlds! 👯‍♂️
```
[`@media (min-width: ${DESKTOP}px)`]: {
...
}
....
const isDesktop = useMediaQuery({ minWidth: DESKTOP })
```

gustaveen replied:
And I think it can be hard to use the same `isDesktop` for all components since the component layout will be more contextual. But I we could definitely go over the breakpoints we have and se if we can find a pattern to standardise a few of them 🙌

styrelius replied:
But maybe you just meant the `maxWidth`? In that case, see comments below (TLDR: I'll remove it).

virrios replied:
What I think could be a problem about useMediaqueries  is that you introduce styling into the component logic. I feel like you should not introduce it where it is maybe not needed (It absolutely has its place sometimes!).  I do however like that it becomes more declarative when the styles should be applied. 
In another project I think we went about it like this, declaring constants for MediaQueries pixelValues, but that might not really add anything to the readability more than just centralising pixelValues we use for mobile etc. 
`
...
@media (max-width: ${MOBILE_WIDTH}px){
`
This is more thoughts in general, I think the pr look super! 💪 


EDIT: Haha I am to slow @gustafguner ... 🐌

styrelius replied:
No sure, you're right - this centralising can be done without using a hook, I was confusing things. I'm changing it to a media query (I still think we should use the same breakpoints everywhere, but this is something we should discuss later and not now).

styrelius replied:
> EDIT: Haha I am to slow @gustafguner ... 🐌

@virrios  Wrong Gustav 😳  But I'm also too slow apparently haha! Anyways, YES valid points both of you!

gustaveen commented on src/client/pages/OfferNew/Checkout/index.tsx:
```diff
@@ -191,6 +183,26 @@ export const Checkout: React.FC<Props> = ({
 
   const outerWrapper = useRef<HTMLDivElement>()
 
+  const [wrapperScrollWidth, setWrapperScrollWidth] = useState<
+    number | undefined
+  >()
+
+  useEffect(() => {
+    const setScrollWidth = () => {
+      if (outerWrapper.current?.scrollWidth) {
+        setWrapperScrollWidth(outerWrapper.current?.scrollWidth)
+      }
+    }
+
+    setScrollWidth()
+
+    window.addEventListener('resize', setScrollWidth)
+
+    return () => {
+      window.removeEventListener('resize', setScrollWidth)
+    }
+  }, [outerWrapper.current?.scrollWidth])
+
```
Didn't fully understand what issue we solved with the scroll width 🤔

styrelius replied:
I tried to explain that in point 4. in the description but I guess I didn't do such a good job then. If the browser has a fixed scrollbar (like Edge and Explorer have, and Google had until like last weekend lol) the `Sign` component will cover that too. Which is ugly. But as I wrote in the description, if you think is just unnecessary complexity that only solves a problem for the few users who use Edge or Explorer (I mean who does, really?) then I can remove it.

styrelius replied:
OK after looking into it more I'm not even sure it matters for Edge so I'll remove it. It was a neat solution though imo, when it was needed for Chrome last week 😄 But I'll remove it!

gustaveen replied:
It was just med that couldn't visualise it my head 😁  It think it's awesome that you took that in to consideration, I'm all for not punishing users with old browsers. But if it's not needed for Edge I agree that we can remove it to not have that complexity 🥳

styrelius commented on src/client/pages/OfferNew/index.tsx:
```diff
@@ -80,9 +78,7 @@ export const OfferNew: React.FC = () => {
-    if (isMobile) {
-      Intercom('update', { hide_default_launcher: open })
-    }
+    Intercom('update', { hide_default_launcher: open })
```
I think you have a point with intercom being accessible from the checkout but unfortunately it's then placed on top of the sign button - since the button is now almost as wide as the sidebar. It looks really weird imo. But maybe we can fins a solution for it that looks better.

styrelius replied:
OK so I guess we're going with just removing the Intercom button for now, since Mange said that's fine? @gustaveen

styrelius committed:
Remove the whole calculating the 'scrollWidth' of 'outerWrapper' ordeal in Checkout component since it's apparently not needed at all anymore. Also remove related code in Sign.tsx.

gustaveen commented:
Awesome work with this! 🥳

styrelius committed:
Add 'DESKTOP_BREAKPOINT' variable in Sign.tsx and use that with a media query in 'Wrapper' styling. Also use it to get the 'isDesktop' value. Bonus: remove redundant 'left: 0' declaration for Sign 'Wrapper' :)

styrelius committed:
Merge branch 'master' into fix/checkout-sign-btn

styrelius committed:
Attempt to fix issue with height of Checkout sidebar in mobile (specifically Safari on iOS):
calculate innerHeight of window and use that as height for the OuterWrapper instead of 100vh.

palmenhq commented on src/client/pages/OfferNew/Checkout/Sign.tsx:
```diff
@@ -1,69 +1,35 @@
+  padding: 1rem;
+  position: sticky;
+  bottom: 0;
+  transition: visibility 3s;
```
i had no idea visibility was animatable! Nice one :D

gustaveen replied:
Me neither! But where do we set the visibility on this component though?

styrelius replied:
Lol it's not - I just accidentally left it there after some experimenting that didn't pan out 😄

palmenhq commented on src/client/pages/OfferNew/index.tsx:
```diff
@@ -97,8 +93,6 @@ export const OfferNew: React.FC = () => {
     )
   }
 
-  // TODO: Add error component to render if there's no offerData
```
lol 😓

styrelius replied:
😅

palmenhq reviewed (dismissed):
Looks awesome imo :) nice refactorings also

gustaveen commented:
I tried on my iPhone X now and the button is hidden and you can't scroll down to it 😬

![ezgif com-gif-maker (1)](https://user-images.githubusercontent.com/6661511/100252453-2af45480-2f40-11eb-9389-d2b4bd249ffc.gif)

I tried quickly with this:
1. `Wrapper` in sign needs to be `position: fixed` to ensure it's visible
2. `Scrollwrapper` needs `-webkit-overflow-scrolling: touch;` to enable scroll on iPhone

It stills seem to scroll the page behind though, so might be something more we need to add back to the `Scrollwrapper` though

styrelius commented:
@gustaveen Yep I know, that's what I've been working on the entire day 😭
![image](https://user-images.githubusercontent.com/42962286/100254702-ee762800-2f42-11eb-843c-42834495b195.png)

I also tried with `position: fixed` but as you say it will scroll the page behind the sidebar then. But hey! Since the sidebar covers the entire screen in mobile I can just set it to `fixed` when it's mobile! Maybe that will solve it

gustaveen commented:
I hear you @styrelius ! It's always a hustle to fix correct scroll behaviour in overlays and such on mobile so no worries 😁  

Regarding the scroll on the content below I think the safest way to fix that is to set `overflow: hidden` on body, that's how material-ui and similar ui libs usually fix it. So one way to add that (might ba a cleaner way but works in my book 🙂 ) is to do it add `document.body.style.overflow = 'hidden'` to `handleCheckoutToggle`

styrelius committed:
Attempt once more to fix height of Checkout sidebar, so that the Sign button won't be hidden in Safari on iOS.
Mainly move the calculated windowHeight prop from OuterWrapper to ScrolLWrapper.

styrelius commented:
OMG I think I [fixed it](https://web-onboardi-fix-checko-fl6zq5.herokuapp.com/no/new-member)!!

@gustaveen Really clever (and kind of dumb in the absolutely best way) to set `overflow: hidden` on body to prevent scroll - I didn't think of that! Still, I didn't use it since we already have this `useScrollLock` hook but that might be a good refactoring for the future. The hook we have now is very specific to the `Checkout` sidebar case and we might want to do this for other components as well.

styrelius committed:
Remove useless 'transition' declaration from Sign Wrapper. Apparently I had forgotten it there.

styrelius committed:
Merge branch 'master' into fix/checkout-sign-btn

gustaveen reviewed (dismissed):
![giphy (1)](https://user-images.githubusercontent.com/6661511/100430325-ecbd7900-3096-11eb-858f-3bac367eca9f.gif)

styrelius committed:
Merge branch 'master' into fix/checkout-sign-btn

styrelius committed:
Render <Sign> outside ScrollWrapper instead of inside of it. Add SignComponentSpacing styled component so that Sign won't cover content in the bottom of InnerWrapper.

gustaveen reviewed (approved):
Boom! 💥
